### DISCUSSION TITLE
Livewire emit event execute multiple times incremental

### DISCUSSION THREAD

**Discussion by hazwan91** [livewire/livewire#5055] [1. Help] [answered]
Hello.

I am trying to resend a verifcation email to my registered users. 

The process is:
1. Click the button (and wire:click will trigger to open modal)
2. A confirmation modal will shown
3. Click yes (using jquery onclick, event will be emitted from javascript to send a verification email)

The problem that i face is that running all the process once will execute once. Then on running the same process next will execute twice and thrice and fourth depending on how many user you will send **without refresh**. In addition, the email that has been sent also doubling, tripling and etc. If after sending mail then i redirect to refresh the page, then try sending another one, all executed once and good.

Relevant Tools:
"livewire/livewire": "^2.10",
"laravel/framework": "^9.11",
"jquery": "^3.6.0"

My layout are followed from the josh hanley tutorial which is located here
[How to structure your layout file for Livewire](https://joshhanley.com.au/articles/how-to-structure-your-layout-file-for-livewire)

As an extra note, i use `Full Page Component`

Thus code would be like this

> app.blade.php
```
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="csrf-token" content="{{ csrf_token() }}">
    {{-- <meta name="turbo-cache-control" content="no-cache"> --}}

    <title>{{ $title ?? config('app.name', 'Laravel') }}</title>

    <link rel="stylesheet" href="{{ asset(mix('css/app.css')) }}">
    @livewireStyles
    {{ $styles ?? null }}
    <script src="{{ asset(mix('js/app.js')) }}" defer></script>
</head>

<body>
    {{ $slot }}
    @livewireScripts
    <script>
        var baseUrl = '{{ config("app.url") }}';
    </script>
    {{ $scripts ?? null }}
</body>

</html>

```

> page.blade.php

```
@auth
<div id="page-container" class="sidebar-o sidebar-inverse enable-page-overlay side-scroll page-header-fixed">
        @include('layouts.authlayout._partials.app_modals')
        @include('layouts.authlayout._partials.right_sidebar')
        @include('layouts.authlayout._partials.sidebar')
        @include('layouts.authlayout._partials.navbar')

        <main id="main-container">
            <div class="content">
                <nav class="breadcrumb push px-2">
                    {{ $breadcrumbs ?? null }}
                </nav>
                {{ $slot }}
            </div>
        </main>
        @include('layouts.authlayout._partials.footer')
    </div>
@endauth

@guest
    <section>
        {{ $slot }}
    </section>
@endguest

```

> auth-user-index.blade.php

```
<x-page-layout>
@if ($user->active === 0)
     <a href="javascript:void(0)" class="btn btn-light btn-sm" data-toggle="tooltip" title="Resend email verification" wire:click.prevent="confirmResendAccountVerificationMail('{{ $user->uuid }}')">
        <i class="si si-envelope"></i>
    </a>
@endif
</x-page-layout>

<x-slot name="scripts">
    <script src="{{ asset(mix('js/auth_user_index.js')) }}"></script>
</x-slot>
```

> auth_user_index.js

```
$(function() {
    window.addEventListener('confirmResendAccountVerificationMail', function(e) {
        window.$('#modal-confirmation-title').text('Send verification email')
        window.$('#modal-confirmation-content').html(`
            <p>Are you sure to send another email verification to this user ?</p>
        `)
        window.$('#modal-confirmation-ok').on('click', function(e) {
            window.livewire.emit('resendAccountVerificationMail')
        })
        window.$('#modal-confirmation').modal('show')
    })
    window.addEventListener('successResendAccountVerificationMail', function(e) {
        window.$('#modal-confirmation').modal('hide')
    })
})

```

> AuthUserIndex.php

```
public $selectedIdForResendAccountVerificationMail = '';

protected $listeners = [
        'resendAccountVerificationMail'
];

public function confirmResendAccountVerificationMail($uuid) {
        $this->selectedIdForResendAccountVerificationMail = $uuid;
        $this->dispatchBrowserEvent('confirmResendAccountVerificationMail');
}

public function resendAccountVerificationMail() {
        dump('here');
        // Mail::to(...).........
}
```

I can see that dump('here') would be execute increasingly many times

I did try to use `@once` on my app.blade.php, page.blade.php and also auth-user-index.blade.php but still the same.

![Capture1](https://user-images.githubusercontent.com/97268905/174949484-00b900bf-15c9-451e-9b5a-4b046cbc7d31.PNG)

![Capture2](https://user-images.githubusercontent.com/97268905/174949497-4be08942-24b4-41aa-a927-2c71a2262001.PNG)

![Capture3](https://user-images.githubusercontent.com/97268905/174949507-264b65a3-c720-4808-bb12-647c2d50edd7.PNG)

Is there something that i fail to understand? Please help.

**Comment by Martleby** [comment]
From your description its sounds like some variety of dom-diffing issue. I suspect what is happening is that on every subsequent request Livewire is appending to your script block, hence  you get multiple event listeners.

Things you could try: add wire:ignore to your script block. Add a wire:key to your script block to help Livewire identify it.

Cheers,
Martin.

**Comment by hazwan91** [comment, marked as answer]
Oh well its working now.. I just realize that my javascript kinda awkward.

Original block of code would be something like this:
```
$(function() {
    window.addEventListener('confirmResendAccountVerificationMail', function(e) {
        window.$('#modal-confirmation-title').text('Send verification email')
        window.$('#modal-confirmation-content').html(`
            <p>Are you sure to send another email verification to this user ?</p>
        `)
        window.$('#modal-confirmation-ok').on('click', function(e) {
            window.livewire.emit('resendAccountVerificationMail')
        })
        window.$('#modal-confirmation').modal('show')
    })
    window.addEventListener('successResendAccountVerificationMail', function(e) {
        window.$('#modal-confirmation').modal('hide')
    })
})
```

Then i change it to something like this:
```
$(function() {
    window.addEventListener('confirmResendAccountVerificationMail', function(e) {
        window.$('#modal-confirmation-title').text('Send verification email')
        window.$('#modal-confirmation-content').html(`
            <p>Are you sure to send another email verification to this user ?</p>
        `)
        window.$('#modal-confirmation').modal('show')
    })
    window.$('#modal-confirmation-ok').on('click', function(e) {
            window.livewire.emit('resendAccountVerificationMail')
    })
    window.addEventListener('successResendAccountVerificationMail', function(e) {
        window.$('#modal-confirmation').modal('hide')
    })
})
```

I switch the $('#modal-confirmation-ok') onclick event button to outside of confirmResendAccountVerificationMail event listener.

I am not that pro with javascript, so i am not able to explain why this is working. Maybe someone can explain this.
